lib/vector-core/src/event/value.rs

```diff
@@ -22,6 +23,46 @@ pub enum Value {
     Null,
 }
 
+impl Eq for Value {}
+
+impl Hash for Value {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        core::mem::discriminant(self).hash(state);
+        match self {
+            Value::Array(v) => {
+                v.hash(state);
+            }
+            Value::Boolean(v) => {
+                v.hash(state);
+            }
+            Value::Bytes(v) => {
+                v.hash(state);
+            }
+            Value::Float(v) => {
+                if v.is_finite() {
+                    let trunc: u64 = unsafe { std::mem::transmute(v.trunc()) };
+                    if trunc == 0 {
+                        v.is_sign_negative().hash(state);
+                    }
+                    trunc.hash(state);
```

Comment:
Could we add a comment to this match arm with how we are doing the hashing? Something like:

```
// This hashes floats with the following rules:
// * NaNs hash as equal
// * -0 and +0 hash to different values
// * otherwise transmute to u64 and hash
```

This one, in particular, makes me think we should have some tests around this hash function. Would you mind adding a few unit tests here?